File: app/javascript/embeds/transistor.js
import MediaSite from "media_site"

class Transistor extends MediaSite {
  constructor() {
    super(
      /https:\/\/share\.transistor\.fm\/[se]\/([a-zA-Z0-9\-]+)(?:\/(latest|playlist|[a-zA-Z0-9\-]+))?\??(.*)$/,

      async (url) => {
        const match = url.match(this.regex)
        if (match) {
          const showSlug = match[1]
          const episodeOrType = match[2] || ''

          if (episodeOrType) {
            return `https://share.transistor.fm/e/${showSlug}/${episodeOrType}`
          } else {
            return `https://share.transistor.fm/e/${showSlug}`
          }
        }
        return null
      },

      (embedUrl) => {
        const iframe = document.createElement('iframe')
        iframe.src = `${embedUrl}?color=444444&background=ffffff`
        iframe.width = "100%"
        iframe.style.border = "none"
        iframe.seamless = true
        iframe.loading = "lazy"

        if (embedUrl.includes('/playlist')) {
          iframe.height = "390"
        } else {
          iframe.height = "180"
        }

        return iframe
      }
    )
  }
}

export default Transistor

File: app/models/media_embed/providers/transistor.rb
module MediaEmbed
  module Providers
    class Transistor < Base
      REGEX = %r{\Ahttps://share\.transistor\.fm/[se]/([a-zA-Z0-9-]+)(?:/(latest|playlist|[a-zA-Z0-9-]+))?}i

      def render(url)
        match = url.match(REGEX)
        return unless match

        show_slug, episode_or_type = match.captures
        embed_url = "https://share.transistor.fm/e/#{show_slug}"
        embed_url += "/#{episode_or_type}" if episode_or_type.present?

        iframe("#{embed_url}?color=444444&background=ffffff",
          width: "100%",
          height: episode_or_type == "playlist" ? "390" : "180",
          seamless: true,
          style: "border: none")
      end
    end
  end
end
